### PR TITLE
EAG2-61 - Start game

### PR DESCRIPTION
**Pull request by RayWayy**

### CODE REVIEW

File: src/main/java/com/eleks/academy/whoami/controller/GameController.java [hunk 1]
```diff
@@ -69,9 +69,8 @@ public ResponseEntity<TurnDetails> findTurnInfo(@PathVariable("id") String id,
 	}
 
 	@PostMapping("/{id}")
-	public ResponseEntity<GameDetails> startGame(@PathVariable("id") String id,
-												 @RequestHeader(PLAYER) String player) {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/controller/GameController.java]: Why did you remove request header, don't forget that to start game can be triggered only by player that is in game, as i see right now anyone can do this if it has ID

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 2]
```diff
@@ -90,65 +89,43 @@ public Map<String, SynchronousPlayer> getPlayers() {
...
 
-		final var nonTakenCharacters = this.suggestedCharacters.values()
-				.stream()
-				.flatMap(Collection::stream)
-				.filter(character -> !character.isTaken())
-				.collect(toList());
+		do {
+			playerToCharacterCopy = getRandomCharacter(playerToCharacterCopy);
+			isValueChanged = isTwoValueEquals(playerToCharacterCopy);
+		} while (!isValueChanged);
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: no need for extra variable you can do as `while (!isTwoValueEquals(playerToCharacterCopy))`

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 3]
```diff
@@ -90,65 +89,43 @@ public Map<String, SynchronousPlayer> getPlayers() {
...
-					nonTakenCharacters.remove(character);
-				});
+		playerCharacter = IntStream.range(0, playerCharacter.size()).boxed()
+				.collect(Collectors.toMap(key::get, value::get));
 
-		return this;
+		return playerCharacter;
 	}
 
+	private boolean isTwoValueEquals(Map<String, String> playerCharacterShuffled) {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: please refactor this method to except two params old and shuffled, to compare them, such a case would work if it is a public method and it compares the value of current to input, but your method is not so, and no need to do it like this. So the best is to have two parameters and work with them

File: src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+		Map<String, String> inputPlayersAndCharacters = new HashMap<>(4);
+		inputPlayersAndCharacters.put("Player1", "Character1");
+		inputPlayersAndCharacters.put("Player2", "Character2");
+		inputPlayersAndCharacters.put("Player3", "Character3");
+		inputPlayersAndCharacters.put("Player4", "Character4");
+
+		var result = characters.assignCharacters();
+
+		assertNotEquals(inputPlayersAndCharacters, result);
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java]: just not equal is not enough you need to check that pairs are not as previous and characters are not duplicated

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]
Also, I don't see a test about a game scenario that will play the game, for now, it has to have all steps before the game start and game start, further it should be modified every time when new aspects of the game will be added

File: src/main/java/com/eleks/academy/whoami/controller/GameController.java (hunk 1 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/controller/GameController.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 2 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 3 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java (hunk 4 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java]: Resolved
I think do correct.

File: src/main/java/com/eleks/academy/whoami/service/impl/GameServiceImpl.java [hunk 5]
```diff
@@ -64,7 +64,7 @@ public void suggestCharacter(String id, String player, CharacterSuggestion sugge
 	}
 
 	@Override
-	public Optional<GameDetails> startGame(String id) {
+	public Optional<GameDetails> startGame(String id, String player) {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/service/impl/GameServiceImpl.java:77]: not just add, you need also to use it

File: src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java [hunk 6]
```diff
@@ -28,7 +31,13 @@ void assignCharactersTest() {
 
 		var result = characters.assignCharacters();
 
-		assertNotEquals(inputPlayersAndCharacters, result);
+		List<String> oldValues = new ArrayList<>(inputPlayersAndCharacters.values());
+		List<String> shuffledValues = new ArrayList<>(result.values());
+
+		for (int i = 0; i < inputPlayersAndCharacters.size(); i++) {
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java]: also don't forget that characters are not duplicated, you need to add this check

File: src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java [hunk 7]
```diff
@@ -28,7 +31,13 @@ void assignCharactersTest() {
 
 		var result = characters.assignCharacters();
 
-		assertNotEquals(inputPlayersAndCharacters, result);
+		List<String> oldValues = new ArrayList<>(inputPlayersAndCharacters.values());
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java]: I would recommend iterating throughout the map, ArrayList doesn't guarantee that element of one list has the same position in another list

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/main/java/com/eleks/academy/whoami/service/impl/GameServiceImpl.java (hunk 5 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/service/impl/GameServiceImpl.java:77]: Resolved

File: src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java (hunk 6 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java (hunk 7 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/core/state/SuggestingCharacterTest.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java [hunk 8]
```diff
@@ -95,7 +94,15 @@ public void answerQuestion(String player, Answer answer) {
...
 	public SynchronousGame start() {
-		return null;
+		var checkState = currentState.peek().getStatus();
+
+		if (checkState.equals(GameStatus.SUGGESTING_CHARACTERS)) {
+			currentState.add(currentState.peek().next());
+			currentState.remove();
+			return this;
+		} else {
+			throw new RuntimeException("Game [" + this.getId() + "] has state [" + this.getStatus() + "]");
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java]: No need of [], write as simple sentence  using string format `format("Game %s has state %s", this.getId(), this.getStatus())`

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 9]
```diff
@@ -59,98 +50,85 @@ public GameStatus getStatus() {
 		return GameStatus.SUGGESTING_CHARACTERS;
 	}
 
-	// TODO: Consider extracting into {@link GameState}
 	private Boolean finished() {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: try to use boolean as primitive to have only two possible option true or false, but not true, false and null

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 10]
```diff
@@ -59,98 +50,85 @@ public GameStatus getStatus() {
...
-		final var enoughCharacters = Optional.of(this.suggestedCharacters)
+		this.toPlayerCharacterMap(this.players);
+
+		final var enoughCharacters = Optional.of(this.playerCharacterMap)
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: much easier just to write `this.playerCharacterMap.values().size()` without any transformations

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 11]
```diff
@@ -59,98 +50,85 @@ public GameStatus getStatus() {
...
 				.map(Map::values)
 				.stream()
 				.mapToLong(Collection::size)
 				.sum() >= this.players.size();
 
-		return this.suggestedCharacters.size() > 1
+		return this.playerCharacterMap.size() > 1
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: why do we need here two conditions? if success can be only and only when a number of suggestions == `players.size()`

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 12]
```diff
@@ -59,98 +50,85 @@ public GameStatus getStatus() {
...
+		Map<String, String> playerCharacterShuffled;
+		int countShuffledCharacters = 0;
 
-		authors.forEach(author -> {
-			final var character = this.getRandomCharacter()
-					.apply(this.suggestedCharacters.get(this.<String>cyclicNext().apply(authors, author)));
+		do {
+			playerCharacterShuffled = shuffledCharacters(playerToCharacterCopy);
+			countShuffledCharacters++;
+			if (countShuffledCharacters == 5) {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: move this 5 into constant and name it, to know what this 5 means

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 13]
```diff
@@ -59,98 +50,85 @@ public GameStatus getStatus() {
...
+				count++;
+			}
+		}
 
-		return this;
+		return count == 0;
 	}
 
+	private static Map<String, String> moveCharacters(Map<String, String> playerCharacter) {
+		NavigableMap<String, String> tmpPlayerCharacter = new TreeMap<>(playerCharacter);
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: It's great that you have implemented your own algorithm, but there is already created Java solution for that, https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#rotate(java.util.List,%20int), also I would like that shift would be random between 1 and `players.size()` and this number would be good to be as a parameter please look `moveCharacters(playerToCharacterCopy, new Random().nextInt(players.size() - 1) + 1)`, and this method also would be good to cover with the test and in test you will test shift by a specific number like `moveCharacters(playerToCharacterCopy, 2)`

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 12 above)
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: I would do this after the loop in case random mixing fails then we do shifting

File: src/test/java/com/eleks/academy/whoami/programTest/GameTest.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+		suggestCharacter(id);
+
+		var startedGame = gameService.startGame(id, player);
+		logger.log(Level.INFO, "Current game details after start: {0}", asJsonStringOptionalGameDetails(startedGame));
+
+		// TODO: create method firstTurn with parameter id for a game after starting
+	}
+
+	private String asJsonString(GameDetails gameDetails) {
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/programTest/GameTest.java]: You may use lombok's `@ToString` annotation over objects that you want to use in log, no need to parse them into JSON there

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/test/java/com/eleks/academy/whoami/programTest/GameTest.java (hunk 14 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/programTest/GameTest.java]: Not resolved
I use parse to JSON because I want to see results pretty. Not in one line. Because fields and information so many.

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 12 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 13 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved
I try to found this method but all time I found for map. For maps not this method.
This is a cool hint.

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 12 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java (hunk 8 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 9 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 10 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 11 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 15]
```diff
@@ -59,98 +49,90 @@ public GameStatus getStatus() {
...
+				countEquals++;
+			}
+		}
 
-		this.players.keySet()
-				.stream()
-				.filter(player -> !authorsSet.contains(player))
-				.forEach(player -> {
-					final var character = this.getRandomCharacter().apply(nonTakenCharacters);
+		return countEquals <= 0;
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: if `int countEquals = 0;` and in the end `countEquals <= 0` then better to compare `countEquals == 0`

in your case method better would sound like `areNotEqual`.

If right now it works then update the condition and update the method name

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java [hunk 16]
```diff
@@ -59,98 +49,90 @@ public GameStatus getStatus() {
...
+	}
 
-					character.markTaken();
+	private Map<String, String> moveCharacters(Map<String, String> playerToCharacterCopy, int randomShiftNumber) {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: this method better rename as `shiftCharactersBy`

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 15 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java (hunk 16 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java]: Resolved

**Review by VolodymyrKr-Eleks — APPROVED** [review]
